crypto: Add support for non-IV-prefixed CBC ciphers in mbedTLS shim

sidcha opened the pull request:
This patch adds a new capability flag `CAP_NO_IV_PREFIX` that allows CBC operation without prefixing the IV to the output ciphertext.

When `CAP_NO_IV_PREFIX` is active, the IV passed to cipher_cbc_op() is not modified. So the length of the ciphertext block is equal to the plaintext block, allowing in-place encryption/decryption.

This flag is useful when the IV is derived from other sources (for instance a chained MAC serving as IV) and not directly transmitted to the other end.

**Notes:**
* Fixes: #20561
* Required for: #19483

tbursztyka commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -103,36 +104,55 @@ int mtls_ecb_decrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
 {
-	int ret;
+	int ret, iv_bytes;
+	u8_t *p_iv, iv_loc[16];
```
I cannot remember why iv is of 16bytes fixed length here, is it because of cbc or because of mbedtls? 

Could you detail? (note: I know next to nothing about crypto)

cbsiddharth replied:
In AES-CBC, IV length is always 1 block size (128 bits) regardless of key lengths (AES-128, AES-256 etc.,) .

tbursztyka replied:
ok

tbursztyka commented on include/crypto/cipher_structs.h:
```diff
@@ -175,6 +175,8 @@ struct cipher_ctx {
 #define CAP_AUTONONCE			BIT(7)
 
+/* Don't prefix IV to cipher blocks */
+#define CAP_NO_IV_PREFIX		BIT(8)
```
you will need to put this api modification into its own commit. Then comes the mbedtls commit 

What about tc_shim commit as well? I wonder how you will be addressing ctr operator where (according to documentation) ivlen = keylen - ctrlen, so is of variable length.

cbsiddharth replied:
> put this api modification into its own commit
OK.

>What about tc_shim
These changes are required for another PR I'm working. I'll get to tc_shim later. Anyway each shim has to advertise its capabilities separately. So I guess it shouldn't be an issue.

>how you will be addressing ctr operator where (according to documentation) ivlen = keylen - ctrlen, so is of variable length
mbdtls_shim currently doesn't support CTR mode. One quick look at the tc_shim suggests that IV isn't prefixed in CTR mode, but need to take a closer look to be sure.

cbsiddharth replied:
Done.

tbursztyka commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -48,11 +48,13 @@ struct mtls_shim_session mtls_sessions[CRYPTO_MAX_SESSION];
 #error "You need to define MBEDTLS_MEMORY_BUFFER_ALLOC_C"
 #endif /* MBEDTLS_MEMORY_BUFFER_ALLOC_C */
 
+#define mtls_get_ctx(c, m) \
```
if you make a macro, make it in capital letters.

cbsiddharth replied:
This is part of the other PR #20434. Will fix there.

tbursztyka replied:
sure

cbsiddharth replied:
@tbursztyka Done. Please re-review.

sidcha committed:
crypto: api: New capability flag CAP_NO_IV_PREFIX

Zephyr crypto API currently does not allow IV to be treated as separate
entity. This is mostly due to fact that underlying libraries expect the
IV to be prefixed to the cipher/plain text for performance reasons. But
there are cases where the IV is derived from other sources and not
directly transmitted to the other end. In such cases, it must be treated
as a first class citizen.

This patch adds a new capability flag `CAP_NO_IV_PREFIX` to the crypto
API that allows operations without prefixing the IV to the cipher/plain
text. When `CAP_NO_IV_PREFIX` is active (and supported), the IV passed
to cipher_*_op() must not be modified.

As a side effect, the length of the cipher/plain texts are equal;
allowing for in-place encryption/decryption.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

cbsiddharth commented:
CI Failure looks unrelated. Can someone please verify?

sidcha committed:
drivers/crypto: mbedTLS_shim add support for CAP_NO_IV_PREFIX

Add support for `CAP_NO_IV_PREFIX` in mbedTLS_shim and advertise this as
one of its capabilities. When this flag is active, the IV passed is
preserved to allow applications to reuse the IV buffer.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

cbsiddharth commented:
@carlescufi, @nashif, @galak please review

cbsiddharth commented:
Ping :)

cbsiddharth commented:
Can someone please review/merge this?

pfalcon commented:
> Can someone please review/merge this?

Trying to act on this request, being someone in the reviewer list, I find that it's somewhat unclear what/why is being done from just commit messages, and requires long tracing thru the linked issues to understand "why". Commit messages is what we'll have in the immediate project history, something we'll start to look at when we get a CVE, etc, etc. So, they rather be clear, especially for security-related patches.

So, if this patch fixes what is effectively a deficiency of Zephyr crypto API, where IV is conflated with input cleartext or output cyphertext, while the common baseline in cryptography is to treat those 3 as clearly distinct entities, then ... - I'd personally prefer the commit messages to say something just like that. What I see now at first glance is that it adds yet another mostly random feature of not very clear purpose, sorry.

cbsiddharth commented:
> I find that it's somewhat unclear what/why is being done from just commit messages

@pfalcon, reading the commit message after so long, I realize how unclear it was. Sorry about that. I've reworded the commit message now, do let me know if this one is okay.

pfalcon commented:
@cbsiddharth: Thanks for the updates. As I mentioned, I'm not familiar with Crypto API, so taking a chance to familiar myself and spot/act on any issues seen around it (e.g. https://github.com/zephyrproject-rtos/zephyr/issues/21820). Also, to clarify, I don't have merge rights, so there will be wait on someone with merge rights. From my side, I'm happy to work towards ensuring that crypto API gets into better shape than it was.

Typo in the new commit message:

> IV to be prefixed to the cipher/pain text for

Should be "plain".

pfalcon commented:
And at the same line:

> cipher/pain text for performnace

Typo in "performance", if you care.

cbsiddharth commented:
> Typo in the new commit message

Fixed. Thanks.

pfalcon commented:
> Fixed. Thanks.

Cool, by I now lost my in-progress batched review comments ;-). This is first time this happens to me. An apparent danger of batching review comments vs posting them immediately. Nevermind of course, just a note to interested parties.

pfalcon commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -103,36 +104,55 @@ int mtls_ecb_decrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
+	u8_t *p_iv, iv_loc[16];
 	mbedtls_aes_context *cbc_ctx = MTLS_GET_CTX(ctx, aes);
 
-	/* Prefix IV to ciphertext */
```
Can we preserve this comment, moving to the appropriate branch, perhaps extending as "Prefix IV to ciphertext, which is default behavior of Zephyr crypto API, unless CAP_NO_IV_PREFIX is requested."

cbsiddharth replied:
Done.

pfalcon replied:
Thanks. Looking forward to a push to approve it then. Let's see if that random CI failure will go away too...

pfalcon commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -103,36 +104,55 @@ int mtls_ecb_decrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
+	}
+
 	ret = mbedtls_aes_crypt_cbc(cbc_ctx, MBEDTLS_AES_DECRYPT, pkt->in_len,
-				    iv, pkt->in_buf + 16, pkt->out_buf);
```
This shows how weird is the old behavior: it allows to pass in arbitrary `iv` pointer, but still assumes there's an IV to skip in the input packet ;-/.

cbsiddharth replied:
I believe that this is embedtls shim's idiosyncrasy. Zephyr's crypto API didn't make any assumptions there.

pfalcon reviewed (approved):
I vouch that I fully understood changes done here and they're good improvement of the Zephyr crypto API indeed. Thanks!

pfalcon commented:
@cbsiddharth: Ok, hopefully with 2 vote-ups this will be merged quicker. Otherwise, I see that you got a few of your patches merged already, so you know merges do happen ;-). One way to speed it up might be to hang out on the Slack channel and ping folks with merge rights. Final way is to bring it up to the TSC meeting again, but that's usually only required for PRs with controversies, definitely not the case here.

cbsiddharth commented:
> so you know merges do happen ;-)

Ohh, I do know they will, eventually. I'm just worried about an upcoming change that's slightly bigger than this one. Don't know how long that's going to take ;-).

pfalcon commented:
>  I'm just worried about an upcoming change that's slightly bigger than this one. Don't know how long that's going to take ;-).

All I can say is that I personally appreciate submitting a small, clear, and standalone changes. In this case, winter holidays definitely affected timing. Big changes are indeed may take longer to process, and IMHO regularly merged by "leap of faith", with common cases of regressions/unexpected behavior detected months later. Oh, btw, do you know that we have tests support in Zephyr tree? ;-)

Good luck with your next patches anyway!

pfalcon commented:
@cbsiddharth, So, https://github.com/zephyrproject-rtos/zephyr/pull/21827 was merged, and there's indeed a conflict now, please rebase.

This PR also fails CI check, and that would be another reason why it's not being merged. Actually, people oftentimes even skip reviewing such, apparently thinking "if it doesn't yet even build, then it's clearly not ready for review". Of course, there can be false positives, but dealing with those effectively falls on the PR author. This PR seems to be unlucky to trigger some random failures  with x86_64 arch. I suggest spending 5 mins to click thru Shippable UI to be able to triage these. One obvious way to restart the build is to rebase your PR. If that doesn't work, you can close and reopen the PR. If you see that some failure is seemingly unrelated, but systematic, you would need to call maintainers of that area for help (git log / CODEOWNERS file in the repo top dir).

cbsiddharth commented:
@pfalcon, thank you for the tips. I'll rebase and sort out the build failure in the morning (I'm traveling now).
